pkg/trace/writer/trace_test.go

```diff
@@ -273,12 +290,91 @@ func TestTraceWriterSyncNoop(t *testing.T) {
 		SynchronousFlushing: false,
 	}
 	t.Run("ok", func(t *testing.T) {
-		tw := NewTraceWriter(cfg)
+		tw := NewTraceWriter(cfg, mockSampler, mockSampler, mockSampler)
 		err := tw.FlushSync()
 		assert.NotNil(t, err)
 	})
 }
 
+func TestTraceWriterAgentPayload(t *testing.T) {
+	srv := newTestServer()
+	defer srv.Close()
+	cfg := &config.AgentConfig{
+		Hostname:   testHostname,
+		DefaultEnv: testEnv,
+		Endpoints: []*config.Endpoint{{
+			APIKey: "123",
+			Host:   srv.URL,
+		}},
+		TraceWriter:         &config.WriterConfig{ConnectionLimit: 200, QueueSize: 40},
+		SynchronousFlushing: true,
+	}
+
+	// helper function to send a chunk to the writer and force a synchronous flush
+	sendRandomSpanAndFlush := func(t *testing.T, tw *TraceWriter) {
+		tw.In <- randomSampledSpans(20, 8)
+		err := tw.FlushSync()
+		assert.Nil(t, err)
+	}
+	// helper function to parse the received payload and inspect the TPS that were filled by the writer
+	assertExpectedTps := func(t *testing.T, priorityTps float64, errorTps float64, rareEnabled bool) {
```

Comment:
nit/question: any particular reason to have a helper (`deserializePayload`) defined outside the test, and others (`sendRandomSpanAndFlush`, `assertExpectedTps`) inside? :)

And for `assertExpectedTps`, any reason to provide context as function arguments in some cases, and by capturing the environment for `srv`?

Reply:
Yes, `deserializePayload` is a generic helper, there is nothing specific to my test. It will allow (although I didn't do it) to extend the test suite.

`sendRandomSpanAndFlush` and `assertExpectedTps` are very specific to my test to make it more readable. If they become relevant to other tests we can extract them, but I didn't anticipate that so it's internal for now.

Reply:
> And for assertExpectedTps, any reason to provide context as function arguments in some cases, and by capturing the environment for srv?

`t` is provided in the `t.Run` callback and in the top level function. TLDR: I capture what I can and for the dynamic parts they come as arguments.